Extensions/ArtifactEngine/Engine/artifactEngine.ts

```diff
@@ -146,12 +194,83 @@ export class ArtifactEngine {
         }
         else {
             this.patternList = artifactEngineOptions.itemPattern.split('\n');
+            if (artifactEngineOptions.enableIncrementalDownload) {
+                this.patternList.push("**\\artifact-metadata.csv");
+            }
         }
     }
 
-    private artifactItemStore: ArtifactItemStore = new ArtifactItemStore();
+    createNewHashMap(sourceProvider: models.IArtifactProvider, itemsToProcess: models.ArtifactItem[], artifactEngineOptions?: ArtifactEngineOptions): Promise<string> {
+        return new Promise((resolve, reject) => {
+            if (!artifactEngineOptions.enableIncrementalDownload) {
+                this.newHashMap = {};
+                resolve();
+            }
+            else {
+                sourceProvider.getArtifactItems(itemsToProcess[0]).then((items: models.ArtifactItem[]) => {
+                    sourceProvider.getArtifactItem(items.find(x => path.normalize(x.path) === path.join(itemsToProcess[0].path, 'artifact-metadata.csv'))).then((hashStream: NodeJS.ReadableStream) => {
+                        var newHashPromise = new Promise((resolve) => {
+                            var newHash = readline.createInterface({
+                                input: hashStream
+                            });
+
+                            newHash.on('line', (line) => {
+                                var words = line.split(',');
+                                this.newHashMap[words[0]] = words[1];
```

Comment:
don't assume CSV content is as expected and handle invalid CSV. we should fall back to doing download without cache in such a case. is this already in place?

Reply:
No check for invalid CSV, but if it is not present then we do normal download.